### PR TITLE
better docker execution

### PR DESCRIPTION
**Pull request by bloodearnest**
- fix: allow exit codes to pass through
- fix: handle possible winpty wrapper in RunFixture
- fix: Add new utils package with run_docker function
- chore: add a test for the jupyter command.
- fix: use run_docker with exec command

### CODE REVIEW

File: opensafely/__init__.py [hunk 1]
```diff
@@ -81,27 +78,25 @@ def add_subcommand(cmd, module):
...
 
     # if `run`ning locally, run `check` in warn mode
     if function == local_run.main and "format-output-for-github" not in kwargs:
         check.main(continue_on_error=True)
 
-    sys.exit(0 if success is not False else 1)
+    # allow functions to return True/False, or an explicit exit code
+    if success is False:
+        exit_code = 1
+    elif success is True:
```
Comment (evansd) [inline opensafely/__init__.py:94]: I think this needs `or success is None` to fully replicate the existing behaviour where functions which don't return anything are assumed successful.

File: opensafely/jupyter.py [hunk 2]
```diff
@@ -57,8 +58,12 @@ def add_arguments(parser):
         default=None,
         help="Port to run on",
     )
-    # opt into looser argument parsing
-    parser.set_defaults(handles_unknown_args=True)
+    parser.add_argument(
+        "jupyter_args",
+        nargs=argparse.REMAINDER,
```
Comment (evansd) [inline opensafely/jupyter.py:63]: This is _much_ nicer!

**Review by evansd — APPROVED** [review]
:ok_hand:

**Comment by bloodearnest** [comment]
Fixes #149

**Pull request merged**